src/psyclone/psyir/symbols/data_type_symbol.py

```diff
@@ -107,6 +107,22 @@ def datatype(self, value):
                 "DataType but got: '{0}'".format(type(value).__name__))
         self._datatype = value
 
+    def copy_properties(self, symbol_in):
+        '''Replace all properties in this object with the properties from
+        symbol_in, apart from the name (which is immutable) and visibility.
+
+        :param symbol_in: the symbol from which the properties are copied.
+        :type symbol_in: :py:class:`psyclone.psyir.symbols.DataSymbol`
+
+        :raises TypeError: if the argument is not the expected type.
+
+        '''
+        if not isinstance(symbol_in, DataTypeSymbol):
+            raise TypeError("Argument should be of type 'DataTypeSymbol' but "
+                            "found '{0}'.".format(type(symbol_in).__name__))
+        super(DataTypeSymbol, self).copy_properties(symbol_in)
+        self._datatype = symbol_in.datatype
```

Comment:
I am not totally sure on this one. If we made some modification inside a deep_copied type, then when generating the code it will not be reflected anywhere because the type is only described in the symbol definition (which in this case is in the external file) and they should match.

Also the INTEGER_TYPE, REAL_TYPE, ... that we have use a single instance of the specific scalar type and re-use the same instance everywhere.. which I think is more akind of a shallow reference for datatype.

Moreover, I see the types do not have a deepcopy method yet. So I would say it shouldn't be deepcopied for now, but probably at some point we need to design/think properly how types should behave as it has been done ad-hoc for now.

Reply:
OK, I think we need a discussion. My feeling was that a type should belong to a Symbol and not be shared but am happy to discuss. Look's like we'll be back to doing so on Teams though :-(

Reply:
(I'm happy to mark this as resolved for now.)